core/trino-main/src/main/java/io/trino/cost/JoinStatsRule.java

```diff
@@ -180,47 +182,41 @@ private PlanNodeStatsEstimate filterByEquiJoinClauses(
-        // clause separately because stats estimates would be way off. Instead we choose so called
-        // "driving clause" which mostly reduces join output rows cardinality and apply UNKNOWN_FILTER_COEFFICIENT
-        // for other (auxiliary) clauses.
-        Queue<EquiJoinClause> remainingClauses = new LinkedList<>(clauses);
-        EquiJoinClause drivingClause = remainingClauses.poll();
-        for (int i = 0; i < clauses.size(); i++) {
-            PlanNodeStatsEstimate estimate = filterByEquiJoinClauses(stats, drivingClause, remainingClauses, session, types);
-            if (result.isOutputRowCountUnknown() || (!estimate.isOutputRowCountUnknown() && estimate.getOutputRowCount() < result.getOutputRowCount())) {
-                result = estimate;
-            }
-            remainingClauses.add(drivingClause);
-            drivingClause = remainingClauses.poll();
-        }
-
-        return result;
-    }
+        // Join equality clauses are usually correlated. Therefore, we shouldn't treat each join equality
+        // clause separately because stats estimates would be way off.
+        List<PlanNodeStatsEstimateInfo> knownEstimates = clauses.stream()
+                .map(clause -> {
+                    ComparisonExpression predicate = new ComparisonExpression(EQUAL, clause.getLeft().toSymbolReference(), clause.getRight().toSymbolReference());
+                    return new PlanNodeStatsEstimateWithClause(filterStatsCalculator.filterStats(stats, predicate, session, types), clause);
+                })
+                .filter(estimateWithClause -> !estimateWithClause.getEstimate().isOutputRowCountUnknown())
+                .collect(toImmutableList());
 
-    private PlanNodeStatsEstimate filterByEquiJoinClauses(
-            PlanNodeStatsEstimate stats,
-            EquiJoinClause drivingClause,
-            Collection<EquiJoinClause> remainingClauses,
-            Session session,
-            TypeProvider types)
-    {
-        ComparisonExpression drivingPredicate = new ComparisonExpression(EQUAL, drivingClause.getLeft().toSymbolReference(), drivingClause.getRight().toSymbolReference());
-        PlanNodeStatsEstimate filteredStats = filterStatsCalculator.filterStats(stats, drivingPredicate, session, types);
-        for (EquiJoinClause clause : remainingClauses) {
-            filteredStats = filterByAuxiliaryClause(filteredStats, clause, types);
+        if (knownEstimates.isEmpty()) {
+            return PlanNodeStatsEstimate.unknown();
         }
-        return filteredStats;
+
+        PlanNodeStatsEstimate combinedEstimate = estimateCorrelatedConjunction(
+                stats,
+                knownEstimates,
+                getJoinMultiClauseIndependenceFactor(session),
+                JoinStatsRule::intersectCorrelatedJoinClause);
+        int unestimatedClausesCount = clauses.size() - knownEstimates.size();
+        return normalizer.normalize(
```

Comment:
mention `TODO` here to apply `UNKNOWN_FILTER_COEFFICIENT` only when default filter factor is on.

Also add a comment to unify how `UNKNOWN_FILTER_COEFFICIENT` is applied for join and filter. In fact, I would just fix it in this PR and move that section to `estimateCorrelatedConjunction` instead of having two similar code copies. I would just apply `UNKNOWN_FILTER_COEFFICIENT` once. I don't think it matters for TPCH/TPC-DS anyway